### PR TITLE
fix C++ compile errors (FreeBSD-current)

### PR DESCRIPTION
**Pull request by cnbatch**
Trying to fix the errors of [Bug 279054 - Compile failure when #include <netlink/netlink_snl.h> in C++ code (edit)](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=279054)

### CODE REVIEW

File: sys/netlink/netlink_snl.h [hunk 1]
```diff
@@ -63,7 +63,7 @@
 
 #define	_NLA_END(_start, _len)	((char *)(_start) + (_len))
 #define NLA_FOREACH(_attr, _start, _len)      \
-        for (_attr = (_start);		\
+        for (_attr = (typeof(_attr))(_start);		\
```
Comment (bsdimp) [inline sys/netlink/netlink_snl.h:66]: Does this work with C still?

File: sys/netlink/netlink_snl.h [hunk 2]
```diff
@@ -104,7 +104,7 @@ lb_allocz(struct linear_buffer *lb, int len)
 		return (NULL);
 	void *data = (void *)(lb->base + lb->offset);
```
Comment (bsdimp) [inline sys/netlink/netlink_snl.h:105]: why not just make this char *?
Or change the return type of this to void *?

File: sys/netlink/netlink_snl.h (hunk 1 above)
Comment (rilysh) [reply sys/netlink/netlink_snl.h:66]: typeof is a compiler extension, it was added in C23.

Comment (bsdimp) [reply sys/netlink/netlink_snl.h:66]: So no then. We don't enable c23 for the kernel at all.

Comment (cnbatch) [reply sys/netlink/netlink_snl.h]: changed to use `__typeof` instead

Comment (rilysh) [reply sys/netlink/netlink_snl.h]: Uh how about doing...
```cpp
#ifdef __cplusplus
# define NLA_FOREACH(_attr, _start, _len)      \
        for (_attr = (decltype(_start))(_start);	\
		((char *)_attr < _NLA_END(_start, _len)) && \
		((char *)NLA_NEXT(_attr) <= _NLA_END(_start, _len));	\
		_attr =  NLA_NEXT(_attr))
#else
# define NLA_FOREACH(_attr, _start, _len)      \
        for (_attr = (_start);		\
		((char *)_attr < _NLA_END(_start, _len)) && \
		((char *)NLA_NEXT(_attr) <= _NLA_END(_start, _len));	\
		_attr =  NLA_NEXT(_attr))
#endif
```

This one would be valid C and C++ code (without using any compiler extensions)... but the only issue (visually) is that two similar macros, one for C and one for C++

File: sys/netlink/netlink_snl.h [hunk 3]
```diff
@@ -102,7 +102,7 @@ lb_allocz(struct linear_buffer *lb, int len)
 	len = roundup2(len, alignof(__max_align_t));
 	if (lb->offset + len > lb->size)
 		return (NULL);
-	void *data = (void *)(lb->base + lb->offset);
+	char *data = (lb->base + lb->offset);
 	lb->offset += len;
 	return ((char *)data);
```
Comment (rilysh) [inline sys/netlink/netlink_snl.h:107]: Since you've used char *data ... you can safely remove this cast

File: sys/netlink/netlink_snl.h (hunk 1 above)
Comment (cnbatch) [reply sys/netlink/netlink_snl.h]: Splitting it into two macros might make future maintenance inconvenience.

How about I change it to `(struct nlattr *)` just like `NLA_NEXT` at line 62?
`NLA_FOREACH` macro uses `NLA_NEXT` internally, and `NLA_NEXT` is directly cast as `(struct nlattr *)`.

Comment (brooksdavis) [reply sys/netlink/netlink_snl.h]: > So no then. We don't enable c23 for the kernel at all.

`typeof()` should be fine in practice. Compiling the following fragment on compiler explorer says clang >= 11 and gcc >= 10 support it with default flags.

```C
int a;
typeof(a) b;
```

File: sys/netlink/netlink_snl.h (hunk 3 above)
Comment (bsdimp) [reply sys/netlink/netlink_snl.h:107]: This looks to be the only active comment, and the only thing that has caught my eye that's in the way of committing.

**Review by bsdimp — APPROVED** [review]
I think this is now good

**Pull request merged**